### PR TITLE
Make the result of SRANDMEMBER random

### LINKED ISSUE #2014: Improve the implementation of SRANDMEMBER
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Motivation

Currently it just returns the first N members, which is not actually random.

We can improve it just like the implementation of HRANDMEMBER.
 (Or if you have better ideas for random, feel free to implement them)

### Solution

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by wsehjk**
This pr is to resolve #2014.
I use **Reservoir Sampling**. When the number we want to select is larger than that of elements in the set, all elements are selected

### CODE REVIEW

File: src/types/redis_set.cc [hunk 1]
```diff
@@ -222,12 +223,26 @@ rocksdb::Status Set::Take(const Slice &user_key, std::vector<std::string> *membe
...
 
+  std::vector<std::string> vec;
+  vec.reserve(count);
   auto iter = util::UniqueIterator(storage_, read_options);
   for (iter->Seek(prefix); iter->Valid() && iter->key().starts_with(prefix); iter->Next()) {
-    InternalKey ikey(iter->key(), storage_->IsSlotIdEncoded());
+    if (n < count) {
+      vec.push_back(iter->key().ToString());
+    } else { // n >= count
+      int random = rand()%n;
```
Comment (mapleFU) [inline src/types/redis_set.cc]: Better use rand in https://en.cppreference.com/w/cpp/numeric/random

Comment (mapleFU) [reply src/types/redis_set.cc:233]: Also, should this be `rand() % count` ?

**Comment by jihuayu** [comment]
> This pr is to solve https://github.com/apache/kvrocks/issues/2014.

Hi @wsehjk, you can use resolve #2014 instead of solve #2014. It will make GitHub auto close issue when PR has been merged.

https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Comment by PragmaTwice** [comment]
Please use `./x.py format` to format your code, otherwise the CI cannot pass.

File: src/types/redis_set.cc [hunk 2]
```diff
@@ -222,12 +223,26 @@ rocksdb::Status Set::Take(const Slice &user_key, std::vector<std::string> *membe
   rocksdb::Slice upper_bound(next_version_prefix);
   read_options.iterate_upper_bound = &upper_bound;
 
+  std::vector<std::string> vec;
+  vec.reserve(count);
```
Comment (PragmaTwice) [inline src/types/redis_set.cc]: The name `vec` here seems meaningless (any vector can be named `vec`). Could you rename it?

File: src/types/redis_set.cc [hunk 3]
```diff
@@ -222,12 +223,29 @@ rocksdb::Status Set::Take(const Slice &user_key, std::vector<std::string> *membe
...
+  std::mt19937 gen(rd());
   auto iter = util::UniqueIterator(storage_, read_options);
   for (iter->Seek(prefix); iter->Valid() && iter->key().starts_with(prefix); iter->Next()) {
-    InternalKey ikey(iter->key(), storage_->IsSlotIdEncoded());
+    ++n;
+    if (n <= count) {
+      iter_keys.push_back(iter->key().ToString());
+    } else {  // n > count
+      std::uniform_int_distribution<> distrib(0, n - 1);
+      int random = distrib(gen);  // [0,n-1)
```
Comment (mapleFU) [inline src/types/redis_set.cc]:
```suggestion
      int random = distrib(gen);  // [0,n-1]
```

**Comment by mapleFU** [comment]
Though a different sampling algorithm is used, general LGTM

File: src/common/status.h [hunk 4]
```diff
@@ -168,7 +168,7 @@ struct StringInStatusOr<T, std::enable_if_t<sizeof(T) < sizeof(std::string)>> :
   StringInStatusOr(StringInStatusOr<U>&& v) : BaseType(new std::string(*std::move(v))) {}  // NOLINT
   template <typename U, typename std::enable_if_t<!StringInStatusOr<U>::inplace, int> = 0>
   StringInStatusOr(StringInStatusOr<U>&& v)  // NOLINT
-      : BaseType((typename StringInStatusOr<U>::BaseType &&)(std::move(v))) {}
+      : BaseType((typename StringInStatusOr<U>::BaseType&&)(std::move(v))) {}
```
Comment (PragmaTwice) [inline src/common/status.h]: Could you revert this change?

**Comment by jihuayu** [comment]
You can enable the GitHub action in your repo (It is free). So that you can find the CI failing issues earlier and more easily.

https://github.com/wsehjk/kvrocks/actions

File: src/common/status.h (hunk 4 above)
Comment (wsehjk) [reply src/common/status.h]: It's the result of `./x.py format`

Comment (git-hulk) [reply src/common/status.h]: @wsehjk The format is introduced by clang-format-16 and Kvrocks is now using clang-format-14, so you can change the clang-format to 14 to make the CI happy.

Comment (wsehjk) [reply src/common/status.h]: Ok, i will reinstall clang-format-14

**Comment by jihuayu** [comment]
It seems like the test run fails. Fix it, victory is within sight.

By the way, you don't need to use force push to make PR only one commit; simply use the Update branch Button is ok
image

**Comment by wsehjk** [comment]
> It seems like the test run fails. Fix it, victory is within sight.
> 
> By the way, you don't need to use force push to make PR only one commit; simply use the Update branch Button is ok image

This is because the maintainer of last repo i have contributed required pr to be one commit, i think it still makes sense here.
Everytime `dev` branch lags behind the `unstable` branch， I run`git pull` and ` git rebase unstable dev` to catch up, then I have to use `force-push`....

**Review by git-hulk — APPROVED** [review]

**Review by mapleFU — APPROVED** [review]
LGTM
